# Data Exploration

## Instructions
There are some data files related to transaction saved under the [data](../data) folder:
- Looking into the data using appropriate functions and extract the fields in the data.
- For each data, describe what the data is about and what fields are saved.

You need to answer the questions and perform the task below:
- How many transactions are in GBP?
- How many transactions are NOT in USD?
- What is the average and mediam transaction in USD?
- Construct a table showing the number of transactions in EACH currency

Note:
- You are NOT ALLOWED to import other library or package
- You can write you own functions
- Your answers should be readable with approprate comments
- You can refer to [markdown cheatsheet](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet) if you are not familar with Markdown

## Import libraries 

In [2]:
# Usual libraries are imported here
import os
import yaml
import dask.dataframe as dd
import pandas as pd
import matplotlib
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## Please perform your tasks below and answer the questions

<font size=4>This part is to describe the Data.</font>

In [10]:
countries = pd.read_csv('../data/countries.csv')
countries.info()
countries.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226 entries, 0 to 225
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   CODE       225 non-null    object
 1   NAME       226 non-null    object
 2   CODE3      226 non-null    object
 3   NUMCODE    226 non-null    int64 
 4   PHONECODE  226 non-null    int64 
dtypes: int64(2), object(3)
memory usage: 9.0+ KB


CODE         NAME CODE3  NUMCODE  PHONECODE
0   AF  Afghanistan   AFG        4         93
1   AL      Albania   ALB        8        355
2   DZ      Algeria   DZA       12        213

1): These are information about countries, it has 226 entries and 5 columns(CODE, NAME, CODE3, NUMCODE, PHONECODE), then print top 3 entries of it.

In [11]:
currency_details = pd.read_csv('../data/currency_details.csv')
currency_details.info()
currency_details.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 184 entries, 0 to 183
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   CCY        184 non-null    object
 1   EXPONENT   184 non-null    int64 
 2   IS_CRYPTO  184 non-null    bool  
dtypes: bool(1), int64(1), object(1)
memory usage: 3.2+ KB


CCY  EXPONENT  IS_CRYPTO
0  AED         2      False
1  AFN         2      False
2  ALL         2      False

2): Information about currency_details, CCY(currency);  EXPONENT(currency exponent);  IS_CRYPTO(crypto currency or not)

In [12]:
fraudsters = pd.read_csv('../data/fraudsters.csv')
fraudsters.info()
fraudsters.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 298 entries, 0 to 297
Data columns (total 1 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   USER_ID  298 non-null    object
dtypes: object(1)
memory usage: 2.5+ KB


USER_ID
0  5270b0f4-2e4a-4ec9-8648-2135312ac1c4
1  848fc1b1-096c-40f7-b04a-1399c469e421
2  27c76eda-e159-4df3-845a-e13f4e28a8b5

3): Information about fraudsters, fraudsters' USER_ID

In [14]:
german_credit_data = pd.read_csv('../data/german_credit_data.csv')
german_credit_data.info()
german_credit_data.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Unnamed: 0        1000 non-null   int64 
 1   Age               1000 non-null   int64 
 2   Sex               1000 non-null   object
 3   Job               1000 non-null   int64 
 4   Housing           1000 non-null   object
 5   Saving accounts   817 non-null    object
 6   Checking account  606 non-null    object
 7   Credit amount     1000 non-null   int64 
 8   Duration          1000 non-null   int64 
 9   Purpose           1000 non-null   object
dtypes: int64(5), object(5)
memory usage: 78.2+ KB


Unnamed: 0  Age     Sex  Job Housing Saving accounts Checking account  \
0           0   67    male    2     own             NaN           little   
1           1   22  female    2     own          little         moderate   
2           2   49    male    1     own          little              NaN   
3           3   45    male    2    free          little           little   
4           4   53    male    2    free          little           little   

   Credit amount  Duration              Purpose  
0           1169         6             radio/TV  
1           5951        48             radio/TV  
2           2096        12            education  
3           7882        42  furniture/equipment  
4           4870        24                  car

4): Information about german_credit_data, it's assessment on people's Age, Sex, Job, Housing, Saving accounts, Checking account, Credit amount, Duration and Purpose

In [17]:
users = pd.read_csv('../data/users.csv')
users.info()
users.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9944 entries, 0 to 9943
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   ID                       9944 non-null   object
 1   HAS_EMAIL                9944 non-null   int64 
 2   PHONE_COUNTRY            9944 non-null   object
 3   IS_FRAUDSTER             9944 non-null   bool  
 4   TERMS_VERSION            8417 non-null   object
 5   CREATED_DATE             9944 non-null   object
 6   STATE                    9944 non-null   object
 7   COUNTRY                  9944 non-null   object
 8   BIRTH_YEAR               9944 non-null   int64 
 9   KYC                      9944 non-null   object
 10  FAILED_SIGN_IN_ATTEMPTS  9944 non-null   int64 
dtypes: bool(1), int64(3), object(7)
memory usage: 786.7+ KB


ID  HAS_EMAIL   PHONE_COUNTRY  \
0  1872820f-e3ac-4c02-bdc7-727897b60043          1  GB||JE||IM||GG   
1  545ff94d-66f8-4bea-b398-84425fb2301e          1  GB||JE||IM||GG   
2  10376f1a-a28a-4885-8daa-c8ca496026bb          1              ES   

   IS_FRAUDSTER TERMS_VERSION                CREATED_DATE   STATE COUNTRY  \
0         False    2018-05-25  2017-08-06 07:33:33.341000  ACTIVE      GB   
1         False    2018-01-01  2017-03-07 10:18:59.427000  ACTIVE      GB   
2         False    2018-09-20  2018-05-31 04:41:24.672000  ACTIVE      ES   

   BIRTH_YEAR     KYC  FAILED_SIGN_IN_ATTEMPTS  
0        1971  PASSED                        0  
1        1982  PASSED                        0  
2        1973  PASSED                        0

5): Information about users, including their ID, whether has email, phone from which country, whether is fraudster, vertion, account created date, state active or others, their country, birth year, KYC state and number of failed sign in

In [22]:
transactions = pd.read_csv('../data/transactions.csv')
transactions.info()
transactions.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 688651 entries, 0 to 688650
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   CURRENCY           688651 non-null  object
 1   AMOUNT             688651 non-null  int64 
 2   STATE              688651 non-null  object
 3   CREATED_DATE       688651 non-null  object
 4   MERCHANT_CATEGORY  223065 non-null  object
 5   MERCHANT_COUNTRY   483055 non-null  object
 6   ENTRY_METHOD       688651 non-null  object
 7   USER_ID            688651 non-null  object
 8   TYPE               688651 non-null  object
 9   SOURCE             688651 non-null  object
 10  ID                 688651 non-null  object
 11  AMOUNT_USD         688651 non-null  int64 
dtypes: int64(2), object(10)
memory usage: 63.0+ MB


CURRENCY  AMOUNT      STATE             CREATED_DATE MERCHANT_CATEGORY  \
0      GBP     175  COMPLETED  2017-12-20 12:46:20.294              cafe   
1      EUR    2593  COMPLETED  2017-12-20 12:38:47.232               bar   
2      EUR    1077  COMPLETED  2017-12-20 12:34:39.668               NaN   
3      GBP     198  COMPLETED  2017-12-20 12:45:50.555       supermarket   
4      EUR     990  COMPLETED  2017-12-20 12:45:32.722               NaN   

  MERCHANT_COUNTRY ENTRY_METHOD                               USER_ID  \
0              GBR         cont  8f99c254-7cf2-4e35-b7e4-53804d42445d   
1              AUS         cont  ed773c34-2b83-4f70-a691-6a7aa1cb9f11   
2              CZE         cont  eb349cc1-e986-4bf4-bb75-72280a7b8680   
3              GBR         cont  dc78fbc4-c936-45d3-a813-e2477ac6d74b   
4              FRA         cont  32958a5c-2532-42f7-94f9-127f2a812a55   

           TYPE SOURCE                                    ID  AMOUNT_USD  
0  CARD_PAYMENT   GAIA  b3332e6f-7865-4d6e-b6a5-370bc75568d8         220  
1  CARD_PAYMENT   GAIA  853d9ff8-a007-40ef-91a2-7d81e29a309a        2885  
2  CARD_PAYMENT   GAIA  04de8238-7828-4e46-91f1-050a9aa7a9df        1198  
3  CARD_PAYMENT   GAIA  2b790b9b-c312-4098-a4b3-4830fc8cda53         249  
4  CARD_PAYMENT   GAIA  6469fc3a-e535-41e9-91b9-acb46d1cc65d        1101

6): Information about transactions, it's each transaction's currency, amount, state, created date, etc.

<font size=4>This part is to answer the Questions.</font>

How many transactions are in GBP?

In [25]:
print(len(transactions[transactions['CURRENCY'] == 'GBP']), 'transactions are in GBP.')

339091 transactions in GBP.


How many transactions are NOT in USD?

In [26]:
print(len(transactions)-len(transactions[transactions['CURRENCY'] == 'USD']), 'transactions are not in USD.')

657109 transactions not in USD.


What is the average and mediam transaction in USD?

In [29]:
print('average', np.mean(transactions[transactions['CURRENCY'] == 'USD']['AMOUNT']), 'transaction in USD.')
print('mediam', np.median(transactions[transactions['CURRENCY'] == 'USD']['AMOUNT']), 'transaction in USD.')

average 11598.75470800837 transaction in USD.
mediam 2000.0 transaction in USD.


Construct a table showing the number of transactions in EACH currency

In [51]:
group = transactions[['CURRENCY', 'AMOUNT']].groupby('CURRENCY').count()
group

AMOUNT
CURRENCY        
AED          847
AUD         2110
BTC          283
CAD         1463
CHF         5761
CZK         1507
DKK         1711
ETH          197
EUR       264695
GBP       339091
HKD          480
HUF         1446
ILS          522
INR          207
JPY          733
LTC          137
MAD          115
NOK         2602
NZD          717
PLN        22362
QAR           28
RON         5837
SEK         1579
SGD          487
THB          690
TRY          338
USD        31542
XRP           38
ZAR         1126